Add Probing Documentation & Best Practices

Beige-Coffee opened the pull request:
This PR is a first-draft submission for Probing & ```ProbabilisticScorer``` documentation (including best practices). Once complete, it will close https://github.com/lightningdevkit/lightningdevkit.org/issues/274.

Looking forward to receiving feedback!

ConorOkus commented on docs/probing.md:
```diff
@@ -0,0 +1,318 @@
+# Probing and Path Finding
+
+## The Challenge of Routing Payments on the Lightning Network
+The Lightning Development Kit (LDK) provides you with the software tools to run a node on the Lightning Network - a collection of tens of thousands of nodes that, together, enable cheap and private Bitcoin transactions. 
```
```suggestion
The Lightning Development Kit (LDK) provides you with the software tools to build a node on the Lightning node. The Lightning Network is a collection of tens of thousands of nodes that, together, enable cheap and private Bitcoin transactions. 
```

ConorOkus commented on docs/probing.md:
````diff
@@ -0,0 +1,318 @@
+1) **Missing Information**: Nodes may not know if a given channel's liquidity is prohibitively unbalanced.
+2) **Path-Finding Optimization**: If multiple payment paths exist, we must decide which path is the best.
+
+To help address both of the above challenges, LDK provides a ```ProbabilisticScorer``` struct. This structure holds information about our node's current view of the Lightning Network graph, including an estimate of available liquidity within each channel. It also records historical liquidity observations, which are updated each time our node either fails or succeeds at making a payment (or probe) ...
````
```suggestion
To help address both of the above challenges, LDK provides a [```ProbabilisticScorer```](https://docs.rs/lightning/latest/lightning/routing/scoring/struct.ProbabilisticScorer.html) struct. This structure holds information about our node's current view of the Lightning Network graph, including an estimate of available liquidity within each channel. It also records historical liquidity observations, which are updated each time our node either fails or succeeds at making a payment (or probe). It's important to note that our node can have multiple ```ProbabilisticScorer``` in use simultaneously. Furthermore, this can actually be quite beneficial, as it allows our node to leverage multiple perspectives of the Lightning Network graph, helping to ensure we don't become too reliant on the same paths when exploring potential payment paths.
```

jkczyz replied:
I think we can drop the part about multiple scorers here. It's more an implementation detail of how to configure probing.

ConorOkus commented on docs/probing.md:
```diff
@@ -0,0 +1,318 @@
+
+Before digging into best practices for probing, it's important to note that popular Lightning Network implementations such as Core Lightning (CLN), Lightning Network Daemon (LND), and Lightning Development Kit (LDK), have different internal representations of the Lightning Network. Therefore, it's not possible to simply "feed" CLN, LND, or LDK the same scoring file or information when setting up  ...
+
+### General Architecture Approaches
```
@jkczyz does this check out? So probing isn't something we would do on a mobile phone?

jkczyz replied:
Yeah, the wallet's vendor would periodically provide an updated scorer for the wallet to use. A mobile wallet would need to be active to send / handle probes. However, a wallet could send a JIT probe before making a payment to help ensure it will make it to the destination. And it could update it's local scorer with recent payment success / failure results, though that information would tend to decay before the next payment, if payments are infrequent.

ConorOkus commented:
Thanks @Beige-Coffee. Wondering if we should at least point to RGS https://lightningdevkit.org/blog/announcing-rapid-gossip-sync/ as way to get a view of the network graph, it's important for mobile users.

jkczyz commented on docs/probing.md:
```diff
@@ -0,0 +1,318 @@
+## The Challenge of Routing Payments on the Lightning Network
+The Lightning Development Kit (LDK) provides you with the software tools to run a node on the Lightning Network - a collection of tens of thousands of nodes that, together, enable cheap and private Bitcoin transactions. 
+
+The challenge is that your node is just one among tens of thousands. While each node has its own internal representation of the Lightning Network Graph, the information within this graph is incomplete to optimally route payments. This is largely because, when nodes announce new (or updated) channels to the network, they do not specify how the capacity is distributed between the two channel partie ...
```
s/Graph/graph

jkczyz commented on docs/probing.md:
````diff
@@ -0,0 +1,318 @@
+
+Probing best practices are discussed in more detail in  [**Best Practices**](#best-practices), however, for now, it's sufficient to note that LDK will update its view of the graph each time real and test payments either succeed or fail. Successes and failures are communicated to our LDK node via the [ScoreUpdate](https://docs.rs/lightning/latest/lightning/routing/scoring/trait.ScoreUpdate.html) t ...
+
+Given that the Lightning Network graph is constantly changing as payments flow throughout the ecosystem, our LDK node must account for the passage of time and assume that channel balances will change even if our node has not probed or otherwise interacted with a channel in a while. To accomplish this, LDK provides ```ProbabilisticScoringDecayParameters```. These parameters configure how each chan ...
````
More of a preference, but consider rw=e-wording the last sentence and linking from the word "docs", for instance, instead of "here".

jkczyz commented on docs/probing.md:
````diff
@@ -0,0 +1,318 @@
+Given that the Lightning Network graph is constantly changing as payments flow throughout the ecosystem, our LDK node must account for the passage of time and assume that channel balances will change even if our node has not probed or otherwise interacted with a channel in a while. To accomplish this, LDK provides ```ProbabilisticScoringDecayParameters```. These parameters configure how each chan ...
+
+### Finding the Optimal Path
+The ```ProbabilisticScorer``` assists in finding the optimal path by providing our node with a perspective of the Lightning Network graph and the estimated available liquidity for each channel. To assist in finding the optimal payment route, LDK provides ```ProbabilisticScoringFeeParameters```. These fee parameters impact how our node optimizes its routing decisions. Broadly speaking, ```Probabil ...
````
For `liquidity_penalty_amount_multiplier_msat`, this was recently changed such that the success probability is based on the total amount but the multiplier is used with the payment amount. See https://github.com/lightningdevkit/rust-lightning/pull/3356.

jkczyz commented on docs/probing.md:
````diff
@@ -0,0 +1,318 @@
+- ```decay_params```: This field holds the decay parameters (```ProbabilisticScoringDecayParameters```) for the scorer. As mentioned above, these are customizable, but LDK also provides defaults. You can learn more about them [here](https://docs.rs/lightning/latest/lightning/routing/scoring/struct.ProbabilisticScoringDecayParameters.html).
+- ```network_graph```: This field holds a reference to the network graph.
+- ```logger```: This field holds a reference to the logger.
+- ```channel_liquidities```: This field holds a ```HashMap```, which maps **short channel IDs** to ```ChannelLiquidity``` objects. These objects include liquidity estimates, historical data for the respective channels, and timestamps indicating when the liquidity estimates were last updated.
````
Given `channel_liquidities` is part of the internal representation, we probably don't need to mention it , unlike the other fields which are passed as parameters.

jkczyz commented on docs/probing.md:
````diff
@@ -0,0 +1,318 @@
+2) **Leverage Your Routing History**: If you record a history of payments forwarded by your node, you can leverage that by identifying large or well-ranked nodes that are 1 or 2 hops beyond frequent next-hops from your node. The goal here is to attempt to identify heavily used paths for payments that route through your node. Of course, you do not know the final destination of each payment, which  ...
+3) **Probe Large Nodes**: As mentioned above, it can be beneficial to probe large nodes on the network, such as ACINQ. This is because large nodes can often be the final destination for payments, so probing these nodes will help your LDK node understand which paths are viable for varying payment amounts. It's also worth noting that since many channels on the Lightning Network are connected to lar ...
+4) **Use Multiple ```ProbabilisticScorer```**: Your node can leverage multiple ```ProbabilisticScorer``` at once. Remember, each scorer has its own internal liquidity estimates and payment history for each channel it knows about. Therefore, it may be beneficial to use a new scorer, which has not been updated, to select a route and then update a different scorer. This approach would help ensure th ...
+5) **Varry The Payment Amount**: Remember, a central part of the path-finding algorithm involves generating penalties for each channel so that your node can select the best path. A key variable when calculating the penalty for each channel is the payment amount. You will likely get different paths for the same source --> destination payment, depending on the amount. Therefore, it's important to v ...
````
s/Varry/Vary
(two places)

jkczyz reviewed (commented):
Need to go through the example still but looks good so far.

jkczyz commented on docs/probing.md:
```diff
@@ -0,0 +1,318 @@
+        it.next().unwrap().0.clone()
+    };
+    
+    // Generate a random amount for the probe - up to 500,000,000 milli-satoshis.
+    let amt = ::rand::random::<u64>() % 500_000_000;
```
It might be worth noting somewhere how a more advance prober might choose an amount. For instance, it could try to hone in on the exact liquidity balance by choosing an amount that falls between the min and max liquidity estimate of a channel. Probing below / above the the min / max wouldn't yield any useful information. Although, changing the amount could result in a different path being chosen. To chose an exact path, they could use `lightning::routing::router::build_route_from_hops`. https://docs.rs/lightning/latest/lightning/routing/router/fn.build_route_from_hops.html

jkczyz commented on docs/probing.md:
````diff
@@ -0,0 +1,318 @@
+
+LDK leverages an event-driven architecture, allowing for asynchronous result notification. Therefore, to update our scorer over time, we'll want to inform our application that it should be on the lookout for payment or probing-related events, as these are what indicate to our node that a given route has succeeded or failed. The specific events that we should be watching for and handling are ```Ev ...
+
+To handle the above events and other tasks that either can or should run in the background, LDK provides the ```process_events_async``` function within the ```lightning-background-processor``` crate. When instantiated, this starts an asynchronous process of handling lightning network node operations, such as maintaining our node's view of the network graph, updating the scorer, and more. 
````
Let's link to the crate docs.

jkczyz commented on docs/probing.md:
```diff
@@ -0,0 +1,318 @@
+let probing_cm = Arc::clone(&channel_manager);
+let probing_graph = Arc::clone(&network_graph);
+let probing_logger = Arc::clone(&logger);
+let probing_scorer = Arc::clone(&scorer);
```
We should use a fresh scorer here rather than the same one used for updates, IIUC.

Beige-Coffee commented:
Thanks, @jkczyz and @ConorOkus for the comments! I'll incorporate both of your feedback over the next few days and let everyone know when it's updated 🫡

Beige-Coffee commented:
@jkczyz @ConorOkus 

I’ve reviewed both of your comments and updated the documentation accordingly. For clarity, I've listed my updates below. FYI, I did not list any simple changes I made (ex: updating or deleting a sentence, adding a link, fixing a typo, etc.).
1. Added “A Note For Mobile Developers” section to “Best Practices” which briefly describes Rapid Gossip Sync and links to the LDK RGS blog.
2. Added a new bullet point to the “Vary The Payment Amount” section in “Best Practices” to discuss how the user can select between the min/max payment and specify the route so that they can better hone their channel balance estimates.
3. Added a fresh scorer to the probing code so that we’re not using the same one we’re updating. This is now consistent with recommendations within the documentation itself.

Happy to continue iterating, if there is more we want to add/change!
